packages/admin-panel-server/src/viz-builder/extractors/validators.ts

```diff
@@ -0,0 +1,57 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { yup } from '@tupaia/utils';
+
+import { VisualisationValidator, DashboardVisualisationObject } from '../types';
+
+export class DraftReportValidator implements VisualisationValidator {
+  validationSchema: yup.ObjectSchema;
+
+  constructor() {
+    this.validationSchema = yup.object().shape({
+      dataElements: yup
+        .array()
+        .of(yup.string())
+        .when('dataGroups', { // dataGroups is required if there are no dataElements 
+          is: (dataGroups: string[]) => !dataGroups || dataGroups.length === 0,
+          then: yup.array().of(yup.string()).required().min(1),
+          otherwise: yup.array().of(yup.string()),
+        }),
+      dataGroups: yup
+        .array()
+        .of(yup.string())
+        .when('dataElements', { // dataElements is required if there are no dataGroups 
+          is: (dataElements: string[]) => !dataElements || dataElements.length === 0,
+          then: yup.array().of(yup.string()).required().min(1),
+          otherwise: yup.array().of(yup.string()),
+        }),
+      aggregations: yup.array(),
+      transform: yup.array(),
+    }, ['dataElements', 'dataGroups']);
+  }
+
+  validate(visualisation: DashboardVisualisationObject) {
+    const { data: dataObject } = visualisation;
+    this.validationSchema.validateSync(dataObject);
+  }
+}
+
+export class DraftDashboardItemValidator implements VisualisationValidator {
```

Comment:
I think some folks may take issue with exporting multiple classes from the same file, however personally I think it's fine in this case 👍